File: tmp/timerirqs.c
// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
// Based on hardirq(8) from BCC by Brendan Gregg.
#include <argp.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <fcntl.h>
#include <bpf/libbpf.h>
#include <bpf/bpf.h>
#include "timerirqs.h"
#include "timerirqs.skel.h"
#include "trace_helpers.h"

struct env {
	bool cpu;
	bool distributed;
	bool nanoseconds;
	time_t interval;
	int times;
	bool timestamp;
	bool verbose;
	char *cgroupspath;
	bool cg;
	int targ_cpu;
} env = {
	.interval = 99999999,
	.times = 99999999,
	.targ_cpu = -1,
};

static volatile bool exiting;

const char *argp_program_version = "timerirqs 0.1";
const char *argp_program_bug_address =
	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
const char argp_program_doc[] =
"Summarize timer irq event time as histograms.\n"
"\n"
"USAGE: timerirqs [--help] [-T] [-N] [-d] [-C] [interval] [count] [-c CG]\n"
"\n"
"EXAMPLES:\n"
"    timerirqs            # sum timer irq event time\n"
"    timerirqs -d         # show timer irq event time as histograms\n"
"    timerirqs 1 10       # print 1 second summaries, 10 times\n"
"    timerirqs -c CG      # Trace process under cgroupsPath CG\n"
"    timerirqs --cpu 1    # only stat irq on cpu 1\n"
"    timerirqs -C         # display separately by CPU\n"
"    timerirqs -NT 1      # 1s summaries, nanoseconds, and timestamps\n";

static const struct argp_option opts[] = {
	{ "CPU", 'C', NULL, 0, "Display separately by CPU", 0 },
	{ "distributed", 'd', NULL, 0, "Show distributions as histograms", 0 },
	{ "cgroup", 'c', "/sys/fs/cgroup/unified", 0, "Trace process in cgroup path", 0 },
	{ "cpu", 's', "CPU", 0, "Only stat irq on selected cpu", 0 },
	{ "timestamp", 'T', NULL, 0, "Include timestamp on output", 0 },
	{ "nanoseconds", 'N', NULL, 0, "Output in nanoseconds", 0 },
	{ "verbose", 'v', NULL, 0, "Verbose debug output", 0 },
	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help", 0 },
	{},
};

static error_t parse_arg(int key, char *arg, struct argp_state *state)
{
	static int pos_args;

	switch (key) {
	case 'h':
		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
		break;
	case 'v':
		env.verbose = true;
		break;
	case 'd':
		env.distributed = true;
		break;
	case 'C':
		env.cpu = true;
		break;
	case 's':
		errno = 0;
		env.targ_cpu = atoi(arg);
		if (errno || env.targ_cpu < 0) {
			fprintf(stderr, "invalid cpu: %s\n", arg);
			argp_usage(state);
		}
		break;
	case 'c':
		env.cgroupspath = arg;
		env.cg = true;
		break;
	case 'N':
		env.nanoseconds = true;
		break;
	case 'T':
		env.timestamp = true;
		break;
	case ARGP_KEY_ARG:
		errno = 0;
		if (pos_args == 0) {
			env.interval = strtol(arg, NULL, 10);
			if (errno) {
				fprintf(stderr, "invalid internal\n");
				argp_usage(state);
			}
		} else if (pos_args == 1) {
			env.times = strtol(arg, NULL, 10);
			if (errno) {
				fprintf(stderr, "invalid times\n");
				argp_usage(state);
			}
		} else {
			fprintf(stderr,
				"unrecognized positional argument: %s\n", arg);
			argp_usage(state);
		}
		pos_args++;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
{
	if (level == LIBBPF_DEBUG && !env.verbose)
		return 0;
	return vfprintf(stderr, format, args);
}

static void sig_handler(int sig)
{
	exiting = true;
}

static int print_map(struct bpf_map *map)
{
	struct irq_key lookup_key = {}, next_key;
	struct info info;
	const char *units = env.nanoseconds ? "nsecs" : "usecs";
    char name[64];
	int fd, err;

	if (!env.distributed) {
		printf("%-33s %11s %6s%5s %6s%5s %7s%5s", "Timerirq", "Total_count",
			"Total_", units, "Avg_", units, "Max_", units);
		if (env.cpu)
			printf(" %6s\n", "CPU");
		else
			printf("\n");
	}

	fd = bpf_map__fd(map);

	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
		err = bpf_map_lookup_elem(fd, &next_key, &info);
		if (err < 0) {
			fprintf(stderr, "failed to lookup infos: %d\n", err);
			return -1;
		}

        switch (next_key.id) {
                case IRQ_ID_LOCAL_TIMER:
                    snprintf(name, sizeof(name), "local_timer");
                break;
            case IRQ_ID_HRTIMER:
                    snprintf(name, sizeof(name), "hrtimer");
                break;
            default:
                fprintf(stderr, "unknown key id: %d\n", next_key.id);
                return -1;
                break;
        }
		if (!env.distributed){
            printf("%-33s %11llu %11llu %12lf %11llu", name, info.count, info.total_time,
                                                    1.*info.total_time/info.count, info.max_time);
			if (env.cpu)
				printf(" %5u", next_key.cpu);
            printf("\n");
        } else {
			if (env.cpu)
				printf("cpu = %u ", next_key.cpu);
			printf("timerirq = %s\n", name);
			print_log2_hist(info.slots, MAX_SLOTS, units);
		}
		lookup_key = next_key;
	}

	memset(&lookup_key, 0, sizeof(lookup_key));

	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
		err = bpf_map_delete_elem(fd, &next_key);
		if (err < 0) {
			fprintf(stderr, "failed to cleanup infos: %d\n", err);
			return -1;
		}
		lookup_key = next_key;
	}

	return 0;
}

int main(int argc, char **argv)
{
	static const struct argp argp = {
		.options = opts,
		.parser = parse_arg,
		.doc = argp_program_doc,
	};
	struct timerirqs_bpf *obj;
	struct tm *tm;
	char ts[32];
	time_t t;
	int err;
	int idx, cg_map_fd;
	int cgfd = -1;

	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (err)
		return err;

	libbpf_set_print(libbpf_print_fn);

	obj = timerirqs_bpf__open();
	if (!obj) {
		fprintf(stderr, "failed to open BPF object\n");
		return 1;
	}

	if (probe_tp_btf("irq_handler_entry")) {
		bpf_program__set_autoload(obj->progs.local_timer_entry, false);
		bpf_program__set_autoload(obj->progs.local_timer_exit, false);
	} else {
		bpf_program__set_autoload(obj->progs.local_timer_entry_btf, false);
		bpf_program__set_autoload(obj->progs.local_timer_exit_btf, false);
	}

    if (probe_tp_btf("hrtimer_expire_entry")) {
        bpf_program__set_autoload(obj->progs.hrtimer_expire_entry, false);
        bpf_program__set_autoload(obj->progs.hrtimer_expire_exit, false);
    } else {
        bpf_program__set_autoload(obj->progs.hrtimer_expire_entry_btf, false);
        bpf_program__set_autoload(obj->progs.hrtimer_expire_exit_btf, false);
    }


	/* initialize global data (filtering options) */
	obj->rodata->filter_cg = env.cg;
	obj->rodata->cpu = env.cpu;
	obj->rodata->targ_cpu = env.targ_cpu;
	obj->rodata->targ_dist = env.distributed;
	obj->rodata->targ_ns = env.nanoseconds;

	err = timerirqs_bpf__load(obj);
	if (err) {
		fprintf(stderr, "failed to load BPF object: %d\n", err);
		goto cleanup;
	}

	/* update cgroup path fd to map */
	if (env.cg) {
		idx = 0;
		cg_map_fd = bpf_map__fd(obj->maps.cgroup_map);
		cgfd = open(env.cgroupspath, O_RDONLY);
		if (cgfd < 0) {
			fprintf(stderr, "Failed opening Cgroup path: %s", env.cgroupspath);
			goto cleanup;
		}
		if (bpf_map_update_elem(cg_map_fd, &idx, &cgfd, BPF_ANY)) {
			fprintf(stderr, "Failed adding target cgroup to map");
			goto cleanup;
		}
	}

	err = timerirqs_bpf__attach(obj);
	if (err) {
		fprintf(stderr, "failed to attach BPF object: %d\n", err);
		goto cleanup;
	}

	signal(SIGINT, sig_handler);

	printf("Tracing timer irq event time... Hit Ctrl-C to end.\n");

	/* main: poll */
	while (1) {
		sleep(env.interval);
		printf("\n");

		if (env.timestamp) {
			time(&t);
			tm = localtime(&t);
			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
			printf("%-8s\n", ts);
		}

		err = print_map(obj->maps.infos);
		if (err)
			break;

		if (exiting || --env.times == 0)
			break;
	}

cleanup:
	timerirqs_bpf__destroy(obj);
	if (cgfd > 0)
		close(cgfd);

	return err != 0;
}

File: tmp/timerirqs.py
#!/usr/bin/env python
# @lint-avoid-python-3-compatibility-imports
#
# timerirqs  Summarize timer IRQ (interrupt) event time.
#            For Linux, uses BCC, eBPF.
#
# USAGE: timerirqs [-h] [-T] [-N] [-C] [-d] [-c CPU] [interval] [outputs]

from __future__ import print_function
from bcc import BPF
from time import sleep, strftime
import argparse
import sys

# arguments
examples = """examples:
    ./timerirqs            # sum timer irq event time
    ./timerirqs -d         # show timer irq event time as histograms
    ./timerirqs 1 10       # print 1 second summaries, 10 times
    ./timerirqs -NT 1      # 1s summaries, nanoseconds, and timestamps
    ./timerirqs -c 1       # sum timer irq event time on CPU 1 only
"""
parser = argparse.ArgumentParser(
    description="Summarize timer irq event time as histograms",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-T", "--timestamp", action="store_true",
    help="include timestamp on output")
parser.add_argument("-N", "--nanoseconds", action="store_true",
    help="output in nanoseconds")
parser.add_argument("-C", "--count", action="store_true",
    help="show event counts instead of timing")
parser.add_argument("-d", "--dist", action="store_true",
    help="show distributions as histograms")
parser.add_argument("-c", "--cpu", type=int,
    help="trace this CPU only")
parser.add_argument("interval", nargs="?", default=99999999,
    help="output interval, in seconds")
parser.add_argument("outputs", nargs="?", default=99999999,
    help="number of outputs")
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)
args = parser.parse_args()
countdown = int(args.outputs)
if args.count and (args.dist or args.nanoseconds):
    print("The --count option can't be used with time-based options")
    exit()
if args.count:
    factor = 1
    label = "count"
elif args.nanoseconds:
    factor = 1
    label = "nsecs"
else:
    factor = 1000
    label = "usecs"
debug = 0

# define BPF program
bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/irq.h>
#include <linux/irqdesc.h>
#include <linux/interrupt.h>

const u32 LOCAL_TIMER = 0;
const u32 HRTIMER = 1;
// Add cpu_id as part of key for irq entry event to handle the case which irq
// is triggered while idle thread(swapper/x, tid=0) for each cpu core.
// Please see more detail at pull request #2804, #3733.
typedef struct entry_key {
    u32 tid;
    u32 cpu_id;
} entry_key_t;

typedef struct hist_key {
    u32 type;
    u64 slot;
} hist_key_t;

BPF_HASH(start, entry_key_t);
// TODO: rename hrtimer_start is conflic
//BPF_HASH(hrtimer_start, entry_key_t);
BPF_HISTOGRAM(dist, hist_key_t);
"""

bpf_text_count = """
# TRACEPOINT_PROBE(irq_vectors, local_timer_entry)
# {
#     struct entry_key key = {};
#     u32 cpu = bpf_get_smp_processor_id();
#
#     FILTER_CPU
#
#     key.tid = bpf_get_current_pid_tgid();
#     key.cpu_id = cpu;
#     return 0;
# }

TRACEPOINT_PROBE(irq_vectors, local_timer_exit)
{
    struct entry_key key = {};
    u32 cpu = bpf_get_smp_processor_id();

    FILTER_CPU

    key.tid = bpf_get_current_pid_tgid();
    key.cpu_id = cpu;

    hist_key_t hist_key = {.slot = 0, .type = LOCAL_TIMER};
    dist.atomic_increment(hist_key);
    return 0;
}
"""

bpf_text_time = """
TRACEPOINT_PROBE(irq_vectors, local_timer_entry)
{
    u64 ts = bpf_ktime_get_ns();
    struct entry_key key = {};
    u32 cpu = bpf_get_smp_processor_id();

    FILTER_CPU

    key.tid = bpf_get_current_pid_tgid();
    key.cpu_id = cpu;

    start.update(&key, &ts);
    return 0;
}

TRACEPOINT_PROBE(irq_vectors, local_timer_exit)
{
    u64 *tsp, delta;
    struct entry_key key = {};
    u32 cpu = bpf_get_smp_processor_id();

    key.tid = bpf_get_current_pid_tgid();
    key.cpu_id = cpu;

    // fetch timestamp and calculate delta
    tsp = start.lookup(&key);
    if (tsp == 0) {
        return 0;   // missed start
    }

    delta = bpf_ktime_get_ns() - *tsp;

    // store as sum or histogram
    hist_key_t hist_key = {.slot = 0, .type = LOCAL_TIMER};
    STORE

    start.delete(&key);
    return 0;
}
"""

if args.count:
    bpf_text += bpf_text_count
else:
    bpf_text += bpf_text_time

# code substitutions
if args.dist:
    bpf_text = bpf_text.replace('STORE',
        'hist_key.slot = bpf_log2l(delta / %d);' % factor +
        'dist.atomic_increment(hist_key);')
else:
    bpf_text = bpf_text.replace('STORE',
        'dist.atomic_increment(hist_key, delta);')
if args.cpu is not None:
    bpf_text = bpf_text.replace('FILTER_CPU',
        'if (cpu != %d) { return 0; }' % int(args.cpu))
else:
    bpf_text = bpf_text.replace('FILTER_CPU', '')
if debug or args.ebpf:
    print(bpf_text)
    if args.ebpf:
        exit()

# load BPF program
b = BPF(text=bpf_text)

if args.count:
    print("Tracing timer irq events... Hit Ctrl-C to end.")
else:
    print("Tracing timer irq event time... Hit Ctrl-C to end.")

# output
exiting = 0 if args.interval else 1
dist = b.get_table("dist")
while (1):
    try:
        sleep(int(args.interval))
    except KeyboardInterrupt:
        exiting = 1

    print()
    if args.timestamp:
        print("%-8s\n" % strftime("%H:%M:%S"), end="")

    if args.dist:
        # dist.print_log2_hist(label, "timerirq", section_print_fn=bytes.decode)
        dist.print_log2_hist(label, "source", lambda v: "local_timer" if v==0 else "hrtimer")
    else:
        print("%-26s %11s" % ("Timer Type","Total " + label))
        for k, v in sorted(dist.items(), key=lambda dist: -dist[1].value):
            print("%-26s %11d" % ("local_timer" if k.type==0 else "hrtimer", v.value / factor))
    dist.clear()

    sys.stdout.flush()

    countdown -= 1
    if exiting or countdown == 0:
        exit()
